File: Shopify/pair_programming/url_shortener.cpp
/*
 * Shopify Pair Programming - URL Shortener
 * Confirmed: Apr 2025 (JoinTaro, Senior SWE Canada), Jan 2026 (PracHub)
 * Also mentioned on Glassdoor (recurring)
 *
 * Problem (Layer 1):
 *   Implement a local URL shortener on a single machine.
 *   shorten(long_url)  -> short_code
 *   resolve(short_code) -> original_url or error
 *   In-memory storage only.
 *
 * Layer 2 — Production Hardening:
 *   - Validate well-formed URLs, reject invalid
 *   - Deduplication: same long URL always returns same short code
 *   - Handle collision in hash-based encoding
 *
 * Layer 3 — Persistence:
 *   - Persist mapping to a JSON file
 *   - Reload on program start (no external DB/Redis)
 *
 * Layer 4 — Discussion:
 *   - Memory growth, what changes when data exceeds one machine?
 *   - How do you handle high read vs. write ratio?
 *
 * Apr 2025 candidate note:
 *   "A full front-to-back solution was expected. Interviewer seemed distracted
 *    but still expected complete working code."
 */

#include <iostream>
#include <string>
#include <unordered_map>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <functional>

// ============================================================
// Layer 1 + 2: In-memory URL shortener with validation and dedup
// ============================================================

class UrlShortener {
public:
    explicit UrlShortener(std::string baseUrl = "https://sho.rt/")
        : baseUrl(std::move(baseUrl)) {}

    std::string shorten(const std::string& longUrl) {
        validateUrl(longUrl);

        // Deduplication: return existing code if URL already shortened
        auto it = longToShort.find(longUrl);
        if (it != longToShort.end()) return baseUrl + it->second;

        std::string code = generateCode(longUrl);
        longToShort[longUrl] = code;
        shortToLong[code] = longUrl;
        return baseUrl + code;
    }

    std::string resolve(const std::string& shortUrl) {
        std::string code = extractCode(shortUrl);
        auto it = shortToLong.find(code);
        if (it == shortToLong.end()) {
            throw std::invalid_argument("Unknown short URL: " + shortUrl);
        }
        return it->second;
    }

    // Layer 3: persist to / load from JSON file
    void saveTo(const std::string& path) const {
        std::ofstream f(path);
        f << "{\n";
        bool first = true;
        for (const auto& [code, url] : shortToLong) {
            if (!first) f << ",\n";
            f << "  \"" << escape(code) << "\": \"" << escape(url) << "\"";
            first = false;
        }
        f << "\n}";
    }

    void loadFrom(const std::string& path) {
        std::ifstream f(path);
        if (!f.is_open()) return; // first run, no file yet

        // Minimal JSON parser for our own format (key:value pairs)
        std::string line;
        while (std::getline(f, line)) {
            auto colon = line.find("\": \"");
            if (colon == std::string::npos) continue;
            std::string code = line.substr(3, colon - 3);
            std::string url  = line.substr(colon + 4);
            if (!url.empty() && url.back() == '"') url.pop_back();
            if (!url.empty() && url.back() == ',') url.pop_back();
            if (!url.empty() && url.back() == '"') url.pop_back();
            shortToLong[code] = url;
            longToShort[url] = code;
        }
    }

private:
    std::string baseUrl;
    std::unordered_map<std::string, std::string> shortToLong;
    std::unordered_map<std::string, std::string> longToShort;

    static void validateUrl(const std::string& url) {
        if (url.size() < 7) throw std::invalid_argument("URL too short: " + url);
        if (url.substr(0, 7) != "http://" && url.substr(0, 8) != "https://") {
            throw std::invalid_argument("URL must start with http:// or https://: " + url);
        }
    }

    std::string generateCode(const std::string& url, int attempt = 0) {
        // Base62 encode a hash of the URL
        static const std::string chars =
            "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        size_t h = std::hash<std::string>{}(url + std::to_string(attempt));
        std::string code;
        for (int i = 0; i < 7; i++) {
            code += chars[h % 62];
            h /= 62;
        }
        // Collision: try again with a different salt
        if (shortToLong.count(code) && shortToLong[code] != url) {
            return generateCode(url, attempt + 1);
        }
        return code;
    }

    std::string extractCode(const std::string& shortUrl) const {
        if (shortUrl.substr(0, baseUrl.size()) == baseUrl) {
            return shortUrl.substr(baseUrl.size());
        }
        return shortUrl; // treat as raw code
    }

    static std::string escape(const std::string& s) {
        std::string out;
        for (char c : s) {
            if (c == '"' || c == '\\') out += '\\';
            out += c;
        }
        return out;
    }
};

// ============================================================
// Layer 4 Discussion Points
// ============================================================
/*
 * Q: Memory growth beyond single machine?
 *    - Shard by hash of long URL -> route to specific node
 *    - Use consistent hashing so adding nodes doesn't rehash everything
 *    - Separate read replicas (reads >> writes for URL shorteners)
 *
 * Q: High read ratio?
 *    - Cache short->long in Redis with CDN in front
 *    - Redirect at edge (Cloudflare Worker) before hitting origin
 *    - Write path: sync to DB; read path: cache-aside with TTL
 *
 * Q: What if the same URL is shortened concurrently on two machines?
 *    - Option A: Accept two codes for same URL (benign, wastes space)
 *    - Option B: Write-through to central store; use SETNX (atomic Redis)
 *    - Shopify context: slight inconsistency acceptable for link shortener;
 *      not acceptable for payment URLs (use SETNX there)
 *
 * Q: Analytics (clicks per short URL)?
 *    - Log redirect events to Kafka, consume async into ClickHouse
 *    - Never block the redirect on analytics write
 */

// ============================================================
// Tests
// ============================================================

void runTests() {
    UrlShortener us;

    // Basic shorten + resolve
    auto short1 = us.shorten("https://example.com/very/long/path?q=1");
    std::cout << "Shortened: " << short1 << std::endl;
    std::cout << "Resolved:  " << us.resolve(short1) << std::endl;

    // Deduplication
    auto short2 = us.shorten("https://example.com/very/long/path?q=1");
    std::cout << "Dedup same: " << (short1 == short2 ? "PASS" : "FAIL") << std::endl;

    // Different URL -> different code
    auto short3 = us.shorten("https://example.com/other");
    std::cout << "Different: " << (short1 != short3 ? "PASS" : "FAIL") << std::endl;

    // Invalid URL
    try {
        us.shorten("not-a-url");
        std::cout << "FAIL: should have thrown" << std::endl;
    } catch (const std::invalid_argument& e) {
        std::cout << "Invalid URL rejected: PASS" << std::endl;
    }

    // Unknown short code
    try {
        us.resolve("https://sho.rt/xxxxxxx");
        std::cout << "FAIL: should have thrown" << std::endl;
    } catch (const std::invalid_argument& e) {
        std::cout << "Unknown code rejected: PASS" << std::endl;
    }

    // Persistence
    us.saveTo("/tmp/urls.json");
    UrlShortener us2;
    us2.loadFrom("/tmp/urls.json");
    std::cout << "Persisted resolve: " << us2.resolve(short1) << std::endl;
}

int main() {
    runTests();
    return 0;
}

File: Shopify/java/UrlShortener.java
import java.util.*;

/**
 * Shopify Pair Programming - URL Shortener
 * Confirmed: Apr 2025 (JoinTaro), Jan 2026 (PracHub), Glassdoor (recurring).
 *
 * Java 17.  Run:  javac UrlShortener.java && java UrlShortener
 *
 * Layer 1: shorten / resolve, in-memory
 * Layer 2: URL validation, dedup, collision handling
 * Layer 3 (discussion): persistence, sharding, CDN edge redirect
 *
 * The store is behind an interface (Dependency Inversion) so the in-memory map
 * can be swapped for a file or DB without touching shortener logic. That is the
 * ONLY abstraction here - resist adding more until a second backend exists (YAGNI).
 */
public class UrlShortener {

    /** SOLID seam: the shortener depends on this, not on a concrete map. */
    interface Store {
        void save(String code, String url);
        Optional<String> findUrl(String code);
        Optional<String> findCode(String url);
    }

    static class InMemoryStore implements Store {
        private final Map<String, String> codeToUrl = new HashMap<>();
        private final Map<String, String> urlToCode = new HashMap<>();
        public void save(String code, String url) { codeToUrl.put(code, url); urlToCode.put(url, code); }
        public Optional<String> findUrl(String code) { return Optional.ofNullable(codeToUrl.get(code)); }
        public Optional<String> findCode(String url)  { return Optional.ofNullable(urlToCode.get(url)); }
    }

    private static final String ALPHABET =
        "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static final int CODE_LENGTH = 7;

    private final String base;
    private final Store store;

    UrlShortener(String base, Store store) { this.base = base; this.store = store; }

    String shorten(String longUrl) {
        validate(longUrl);
        // Dedup: the same URL always maps to the same code.
        Optional<String> existing = store.findCode(longUrl);
        if (existing.isPresent()) return base + existing.get();

        String code = generateCode(longUrl, 0);
        store.save(code, longUrl);
        return base + code;
    }

    String resolve(String shortUrl) {
        String code = shortUrl.startsWith(base) ? shortUrl.substring(base.length()) : shortUrl;
        return store.findUrl(code)
            .orElseThrow(() -> new NoSuchElementException("Unknown short URL: " + shortUrl));
    }

    private void validate(String url) {
        if (url == null || !(url.startsWith("http://") || url.startsWith("https://")))
            throw new IllegalArgumentException("URL must start with http:// or https://: " + url);
    }

    private String generateCode(String url, int attempt) {
        // toUnsignedLong avoids the Math.abs(Integer.MIN_VALUE) trap (stays negative!).
        long h = Integer.toUnsignedLong((url + "#" + attempt).hashCode());
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < CODE_LENGTH; i++) { sb.append(ALPHABET.charAt((int)(h % 62))); h /= 62; }
        String code = sb.toString();
        // Collision: a different URL hashed to the same code -> re-salt and retry.
        Optional<String> owner = store.findUrl(code);
        if (owner.isPresent() && !owner.get().equals(url)) return generateCode(url, attempt + 1);
        return code;
    }

    /* ---- Layer 3 talking points ----
     * Persistence: write-through to the Store; on boot, hydrate the in-memory index.
     *   (Don't hand-roll a JSON parser in production - use Jackson; here, a DB row.)
     * Scale beyond one box: shard by hash(longUrl) with consistent hashing.
     *   Reads >> writes -> cache code->url in Redis/CDN, redirect at the edge.
     * Concurrency: two nodes shortening the same URL -> accept two codes (benign)
     *   OR use SETNX for a single canonical code when it matters (paid links).
     * Analytics: emit click events to Kafka; never block the redirect on a write.
     */

    public static void main(String[] args) {
        UrlShortener us = new UrlShortener("https://sho.rt/", new InMemoryStore());

        String s1 = us.shorten("https://example.com/very/long/path?q=1");
        check("resolve",       us.resolve(s1).equals("https://example.com/very/long/path?q=1"));

        String s2 = us.shorten("https://example.com/very/long/path?q=1");
        check("dedup",         s1.equals(s2));

        String s3 = us.shorten("https://example.com/other");
        check("distinct",      !s1.equals(s3));

        try { us.shorten("not-a-url");                 check("reject invalid", false); }
        catch (IllegalArgumentException e)           { check("reject invalid", true); }

        try { us.resolve("https://sho.rt/zzzzzzz");    check("reject unknown", false); }
        catch (NoSuchElementException e)             { check("reject unknown", true); }
    }

    static void check(String name, boolean pass) {
        System.out.printf("%-16s %s%n", name, pass ? "PASS" : "FAIL");
    }
}
